### PR TITLE
`azurerm_mssql_database` / `azurerm_mssql_server` - Added fields for database-level TDE

### LINKED ISSUE #24253: mssql_database creation fails when there is a policy which checks transparent data encryption
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Terraform Version

1.3.10

### AzureRM Provider Version

3.82.0+

### Affected Resource(s)/Data Source(s)

azurerm_mssql_database

### Terraform Configuration Files

```hcl
resource "azurerm_mssql_database" "example" {
  name           = "example-db"
  server_id      = azurerm_mssql_server.example.id
  collation      = "SQL_Latin1_General_CP1_CI_AS"
  license_type   = "LicenseIncluded"
  max_size_gb    = 4
  read_scale     = true
  sku_name       = "S0"
  zone_redundant = true
  enclave_type   = "VBS"

  transparent_data_encryption_enabled = true
}
```

### Debug Output/Panic Output

```shell
"While enabling Transparent Data Encryption... Unexpected status 403: Request disallowed by policy..."
```

### Steps to Reproduce

We have an Azure Policy which prevents MSSQL databases from being created without TDE enabled.
As I noticed from the debug log, `azurerm` creates an MSSQL database sending 2 PUT requests:
- one for the database itself
- second for enabling TDE
The Azure policy blocks the database immediately preventing the 2nd request from finishing successfully.

Is it possible to change the behaviour and create a database with TDE in one request?

### Important Factoids

This issue is introduced from version 3.82.0 after https://github.com/hashicorp/terraform-provider-azurerm/pull/23721

### Workaround

User azurerm version =< 3.81.0

### LINKED ISSUE #24316: TDE with database level customer-managed keys - using TerraformSupport for [thing]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Description

I am trying to create my database in azure with TDE with database level customer-managed keys - using Terraform. Need help in achieving that.

I see --encryption-protector $keyid can be used in creation of db using Azure CLI. Is there something similar to use in terraform?

We have azurerm_mssql_server_transparent_data_encryption" for TDE at server level. I am looking for something similar for database level.

### New or Affected Resource(s)/Data Source(s)

azurerm_mssql_database

### Potential Terraform Configuration

```hcl
resource "azurerm_mssql_database_transparent_data_encryption" "msql_db_tde"{
  server_id        = azurerm_mssql_server.sql_server.id
  database_name = azurerm_mssql_database.databases.name
  key_vault_key_id = data.azurerm_key_vault_key.mssql_db_key.id
}
```

### References

Reference:
https://learn.microsoft.com/en-us/azure/azure-sql/database/transparent-data-encryption-byok-database-level-basic-actions?view=azuresql&tabs=azure-portal%2Cazure-portal2#view-the-database-level-customer-managed-key-settings-on-an-azure-sql-database

### PR DESCRIPTION
**Pull request by dkuzmenok**
Changes
======

- Fixed issue with setting `transparent_data_encryption_enabled`, when it is already set to same value. It was causing policy failures (looks like Setting `Enabled` to already `Enabled` was causing `Disable -> Enable` chain in new API, so added a check to prevent it).
- In data source `azurerm_mssql_database` new fields: `identity`, `transparent_data_encryption_enabled`, `transparent_data_encryption_key_vault_key_id`, `auto_key_rotation_enabled`
- In data source `azurerm_mssql_server` new fields: `transparent_data_encryption_key_vault_key_id`
- In resource `azurerm_mssql_database` new fields: `identity`, `transparent_data_encryption_key_vault_key_id`, `auto_key_rotation_enabled`

Basically we add support for setting TDE on a database level, that is supported with a newer API we now use.
Tests passed on my side.

Fixes #24316 
Fixes #24253

### CODE REVIEW

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks for the PR @dkuzmenok - but it looks like we have a test failure
```
------- Stdout: -------
=== RUN   TestAccMsSqlDatabase_complete
=== PAUSE TestAccMsSqlDatabase_complete
=== CONT  TestAccMsSqlDatabase_complete
    testcase.go:113: Step 3/4 error: Error running apply: exit status 1
        
        Error: unable to parse key: "": cannot parse azure key vault child ID: parse "": empty url
        
          with azurerm_mssql_database.test,
          on terraform_plugin_test.tf line 37, in resource "azurerm_mssql_database" "test":
          37: resource "azurerm_mssql_database" "test" {
        
--- FAIL: TestAccMsSqlDatabase_complete (564.94s)
FAIL
```

**Comment by dkuzmenok** [comment]
> Thanks for the PR @dkuzmenok - but it looks like we have a test failure
> 
> ```
> ------- Stdout: -------
> === RUN   TestAccMsSqlDatabase_complete
> === PAUSE TestAccMsSqlDatabase_complete
> === CONT  TestAccMsSqlDatabase_complete
>     testcase.go:113: Step 3/4 error: Error running apply: exit status 1
>         
>         Error: unable to parse key: "": cannot parse azure key vault child ID: parse "": empty url
>         
>           with azurerm_mssql_database.test,
>           on terraform_plugin_test.tf line 37, in resource "azurerm_mssql_database" "test":
>           37: resource "azurerm_mssql_database" "test" {
>         
> --- FAIL: TestAccMsSqlDatabase_complete (564.94s)
> FAIL
> ```

Please, make another run after my update to test cases.

**Comment by WodansSon** [comment]
The two failures are due to API internal server errors and the other failure is in VM's so not related to this PR.

image

File: internal/services/mssql/mssql_database_data_source.go [hunk 1]
```diff
@@ -90,13 +92,31 @@ func dataSourceMsSqlDatabase() *pluginsdk.Resource {
...
+				Type:     pluginsdk.TypeBool,
+				Computed: true,
+			},
+
+			"transparent_data_encryption_key_vault_key_id": {
+				Type:     pluginsdk.TypeString,
+				Computed: true,
+			},
+
+			"auto_key_rotation_enabled": {
```
Comment (katbyte) [inline internal/services/mssql/mssql_database_data_source.go:107]:
```suggestion
			"transparent_data_encryption_key_automatic_rotation_enabled": {
```

**Review by katbyte — CHANGES_REQUESTED** [review]
thanks for fixing the test @dkuzmenok - aside from one comment about a property name i think this looks good

File: internal/services/mssql/mssql_database_data_source.go (hunk 1 above)
Comment (dkuzmenok) [reply internal/services/mssql/mssql_database_data_source.go]: @katbyte Changed the name now.

**Comment by WodansSon** [comment]
Both test fail with internal server errors, API issue... LGTM! 🚀 

image
